File: examples-nosql-java-sdk/sqlexamples/TableJoins.java
/*Copyright (c) 2020, 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
 */
import java.io.File;
import oracle.nosql.driver.NoSQLHandle;
import oracle.nosql.driver.NoSQLHandleConfig;
import oracle.nosql.driver.NoSQLHandleFactory;
import oracle.nosql.driver.Region;
import oracle.nosql.driver.iam.SignatureProvider;
import oracle.nosql.driver.kv.StoreAccessTokenProvider;
import oracle.nosql.driver.ops.QueryRequest;
import oracle.nosql.driver.ops.QueryResult;
import oracle.nosql.driver.ops.GetRequest;
import oracle.nosql.driver.ops.GetResult;
import oracle.nosql.driver.ops.PutRequest;
import oracle.nosql.driver.ops.PutResult;
import oracle.nosql.driver.ops.TableLimits;
import oracle.nosql.driver.ops.TableRequest;
import oracle.nosql.driver.ops.TableResult;
import oracle.nosql.driver.values.MapValue;
import oracle.nosql.driver.ops.QueryIterableResult;

public class TableJoins{
   /* Name of your tables */
   final static String tableName = "ticket";
   final static String childTableName = "bagInfo";
   final static String fullChildTableName = "ticket.bagInfo";
   final static String descTableName = "flightLegs";
   final static String fullDescTableName = "ticket.bagInfo.flightLegs";

   public static void main(String[] args) throws Exception {
      /*Uncomment the lines of code below if you are using Oracle NoSQL Database Cloud service
      Add the appropriate values of your region and compartment OCID*/
      /*String region ="<your_region_identifier>";
      String compId ="<ocid_of_your_compartment>";
      NoSQLHandle handle = generateNoSQLHandleCloud(region,compId);*/

      /*Uncomment the 2 lines of code below if you are using onPremise Oracle NoSQL Database
      give appropriate value of your endpoint for the onPremise kvstore.*/
      /*String kvstore_endpoint ="http://phoenix71003.dev3sub1phx.databasede3phx.oraclevcn.com:8080";
      NoSQLHandle handle = generateNoSQLHandleonPrem(kvstore_endpoint);*/
      try {
         String crttab_stmt = "CREATE TABLE IF NOT EXISTS " + tableName +
                                                                 "(ticketNo LONG," +
                                                                 "confNo STRING," +
                                                                 "PRIMARY KEY(ticketNo))";
         createTable(handle,crttab_stmt,true,tableName);
         String crtchildtab_stmt = "CREATE TABLE IF NOT EXISTS " + fullChildTableName +
                                                                 "(id LONG," +
                                                                 "tagNum LONG," +
                                                                 "routing STRING," +
                                                                 "lastActionCode STRING," +
                                                                 "lastActionDesc STRING," +
                                                                 "lastSeenStation STRING," +
                                                                 "lastSeenTimeGmt TIMESTAMP(4)," +
                                                                 "bagArrivalDate TIMESTAMP(4)," +
                                                                 "PRIMARY KEY(id))";
         createTable(handle,crtchildtab_stmt,false,childTableName);
         String crtdesctab_stmt = "CREATE TABLE IF NOT EXISTS " + fullDescTableName +
                                                              "(flightNo STRING," +
                                                              "flightDate TIMESTAMP(4)," +
                                                              "fltRouteSrc STRING," +
                                                              "fltRouteDest STRING," +
                                                              "estimatedArrival TIMESTAMP(4)," +
                                                              "actions JSON," +
                                                              "PRIMARY KEY(flightNo))";
         createTable(handle,crtdesctab_stmt,false,descTableName);
         String data1 ="{\"ticketNo\": \"1762344493810\","+
                         "\"confNo\" : \"LE6J4Z\"}";
         writeRowData(handle,"ticket",data1);
         String data2 ="{\"ticketNo\":\"1762344493810\","+
                          "\"id\":\"79039899165297\","+
                          "\"tagNum\":\"17657806255240\","+
                          "\"routing\":\"MIA/LAX/MEL\","+
                          "\"lastActionCode\":\"OFFLOAD\","+
                          "\"lastActionDesc\":\"OFFLOAD\","+
                          "\"lastSeenStation\":\"MEL\","+
                          "\"lastSeenTimeGmt\":\"2019-02-01T16:13:00Z\","+
                          "\"bagArrivalDate\":\"2019-02-01T16:13:00Z\""+
         "}";
         writeRowData(handle,"ticket.bagInfo",data2);
         String data3 ="{\"ticketNo\":\"1762344493810\","+
                        "\"id\":\"79039899165297\","+
                        "\"flightNo\":\"BM604\","+
                        "\"flightDate\":\"2019-02-01T06:00:00Z\","+
                        "\"fltRouteSrc\":\"MIA\","+
                        "\"fltRouteDest\":\"LAX\","+
                        "\"estimatedArrival\":\"2019-02-01T11:00:00Z\","+
                        "\"actions\":[ { "+
                           "\"actionAt\" : \"MIA\","+
                           "\"actionCode\" : \"ONLOAD to LAX\","+
                           "\"actionTime\" : \"2019-02-01T06:13:00Z\""+
                           "}, {" +
                           "\"actionAt\" : \"MIA\","+
                           "\"actionCode\" : \"BagTag Scan at MIA\","+
                           "\"actionTime\" : \"2019-02-01T05:47:00Z\""+
                           "}, {" +
                           "\"actionAt\" : \"MIA\","+
                           "\"actionCode\" : \"Checkin at MIA\","+
                           "\"actionTime\" : \"2019-02-01T04:38:00Z\""+
                        "} ]"+
                     "}";
         writeRowData(handle,"ticket.bagInfo.flightLegs",data3);
         fetchRows(handle);
      } catch (Exception e) {
         System.err.print(e);
      } finally {
         handle.close();
      }
   }
   /* Create a NoSQL handle to access the onPremise Oracle NoSQL database */
   private static NoSQLHandle generateNoSQLHandleonPrem(String kvstore_endpoint) throws Exception {
      NoSQLHandleConfig config = new NoSQLHandleConfig(kvstore_endpoint);
      config.setAuthorizationProvider(new StoreAccessTokenProvider());
      /* If using a secure store uncomment the line below and pass the username, password of the store to StoreAccessTokenProvider*/
      /*config.setAuthorizationProvider(new StoreAccessTokenProvider(username, password));*/
      NoSQLHandle handle = NoSQLHandleFactory.createNoSQLHandle(config);
      return handle;
   }
   /* Create a NoSQL handle to access the cloud service */
   private static NoSQLHandle generateNoSQLHandleCloud(String region, String compId) throws Exception {
      SignatureProvider ap = new SignatureProvider();
      NoSQLHandleConfig config = new NoSQLHandleConfig(region, ap);
      /* set your default compartment*/
      config.setDefaultCompartment(compId);
      NoSQLHandle handle = NoSQLHandleFactory.createNoSQLHandle(config);
      return handle;
   }
   /**
     * Creates a table and sets your desired table capacity
     */
   private static void createTable(NoSQLHandle handle, String sql_stmt,boolean reg_table, String tableName) throws Exception {
      TableRequest treq = new TableRequest().setStatement(sql_stmt);
      if (reg_table== true){
         TableLimits limits = new TableLimits(20, 20, 1);
         treq = new TableRequest().setStatement(sql_stmt).setTableLimits(limits);
      }
      TableResult tres = handle.tableRequest(treq);
      /* The request is async,
       * so wait for the table to become active.
       */
      tres.waitForCompletion(handle, 60000, /* wait 60 sec */
            1000); /* delay ms for poll */
      System.out.println("Table " + tableName + " is active");
   }

   /**
     * Make a row in the table and write it
     */
   private static void writeRowData(NoSQLHandle handle,String table_name,String row_data) throws Exception {
      PutRequest putRequest =
         new PutRequest().setValueFromJson(row_data,null).setTableName(table_name);
      PutResult putResult = handle.put(putRequest);
      if (putResult.getVersion() != null) {
         System.out.println("Wrote " + row_data);
      } else {
         System.out.println("Put failed");
      }
   }
   /* fetch rows based on joins*/
   private static void fetchRows(NoSQLHandle handle) throws Exception {

      try (
         QueryRequest queryRequest = new QueryRequest().setStatement("SELECT * FROM ticket a LEFT OUTER JOIN ticket.bagInfo.flightLegs b ON a.ticketNo=b.ticketNo");
         QueryIterableResult results = handle.queryIterable(queryRequest)) {
            System.out.println("Query results:");
            for (MapValue res : results) {
               System.out.println("\t" + res);
            }
         }
   }
}

File: examples-nosql-python-sdk/sqlexamples/TableJoins.py
# Copyright (c) 2023, 2024 Oracle and/or its affiliates.
# Licensed under the Universal Permissive License v 1.0 as shown at
# https://oss.oracle.com/licenses/upl/
import os
from borneo import (Regions, NoSQLHandle, NoSQLHandleConfig, PutRequest,QueryRequest,
                    TableRequest, GetRequest, TableLimits, State)
from borneo.iam import SignatureProvider
from borneo.kv import StoreAccessTokenProvider

# Given a region, and compartment, instantiate a connection to the
# cloud service and return it
def get_connection_cloud():
   print("Connecting to the Oracle NoSQL Cloud Service")
   #replace the placeholder with your region identifier
   region = '<your_region_identifier>'
   provider = SignatureProvider()
   #If using the DEFAULT profile with the config file in default location
   # ~/.oci/config
   config = NoSQLHandleConfig(region, provider)
   #replace the placeholder with the ocid of your compartment
   config.set_default_compartment("<ocid_of_your_compartment>")
   return(NoSQLHandle(config))

# Given a endpoint, instantiate a connection to onPremise Oracle NoSQL database
def get_connection_onprem():
   #replace the placeholder with the name of your host
   kvstore_endpoint ='http://<hostname>:8080'
   provider = StoreAccessTokenProvider()
   #If using a secure store, uncomment the line below and pass the username,
   # password of the store to StoreAccessTokenProvider
   #provider = StoreAccessTokenProvider(username, password)
   return NoSQLHandle(NoSQLHandleConfig(kvstore_endpoint, provider))

# Create a table and set the table limits
def create_table(handle,statement,table_flag,table_name):
   if (table_flag == 'true'):
      request = TableRequest().set_statement(statement).set_table_limits(TableLimits(20, 10, 1))
   else:
      request = TableRequest().set_statement(statement)

   # Create the table, waiting for a total of 40000 milliseconds
   # and polling the service every 3000 milliseconds to see if table is active
   table_result = handle.do_table_request(request, 40000, 3000)
   table_result.wait_for_completion(handle, 40000, 3000)
   if (table_result.get_state() == State.ACTIVE):
      print('Created table: ',table_name)
   else:
      raise NameError('Table is in an unexpected state ' + str(table_result.get_state()))

def insert_record(handle,table_name,tbl_data):
   request = PutRequest().set_table_name(table_name).set_value_from_json(tbl_data)
   handle.put(request)
   print('Loaded a row into table:',table_name)

# Fetch data from the table based on joins
def fetch_data(handle,sqlstmt):
   request = QueryRequest().set_statement(sqlstmt)
   print('Query results for: ' + sqlstmt)
   result = handle.query(request)
   for r in result.get_results():
      print('\t' + str(r))

def main():
   handle = None
   # if cloud service uncomment this
   handle = get_connection_cloud()
   # if onPremise uncomment this
   # handle = get_connection_onprem()
   tbl_name='ticket'
   regtbl_crtstmt = '''create table if not exists ticket (ticketNo LONG,
                                                          confNo STRING,
                                                          PRIMARY KEY(ticketNo))'''
   childtbl_name='bagInfo'
   childtbl_crtstmt = '''create table if not exists ticket.bagInfo (id LONG,
                                                                    tagNum LONG,
                                                                    routing STRING,
                                                                    lastActionCode STRING,
                                                                    lastActionDesc STRING,
                                                                    lastSeenStation STRING,
                                                                    lastSeenTimeGmt TIMESTAMP(4),
                                                                    bagArrivalDate TIMESTAMP(4),
                                                                    PRIMARY KEY(id))'''
   desctbl_crtstmt = '''create table if not exists ticket.bagInfo.flightLegs (flightNo STRING,
                                                                               flightDate TIMESTAMP(4),
                                                                               fltRouteSrc STRING,
                                                                               fltRouteDest STRING,
                                                                               estimatedArrival TIMESTAMP(4),
                                                                               actions JSON,
                                                                               PRIMARY KEY(flightNo))'''
   desctbl_name='flightLegs'
   create_table(handle,regtbl_crtstmt,'true',tbl_name)
   create_table(handle,childtbl_crtstmt,'false',childtbl_name)
   create_table(handle,desctbl_crtstmt,'false',desctbl_name)
   data1='''{
      "ticketNo": "1762344493810",
      "confNo" : "LE6J4Z"
   }'''
   data2='''{
      "ticketNo": "1762344493810",
      "id":"79039899165297",
      "tagNum":"17657806255240",
      "routing":"MIA/LAX/MEL",
      "lastActionCode":"OFFLOAD",
      "lastActionDesc":"OFFLOAD",
      "lastSeenStation":"MEL",
      "lastSeenTimeGmt":"2019-02-01T16:13:00Z",
      "bagArrivalDate":"2019-02-01T16:13:00Z"
   }'''
   data3='''{
      "ticketNo":"1762344493810",
      "id":"79039899165297",
      "flightNo":"BM604",
      "flightDate":"2019-02-01T06:00:00Z",
      "fltRouteSrc":"MIA",
      "fltRouteDest":"LAX",
      "estimatedArrival":"2019-02-01T11:00:00Z",
      "actions":[ {
         "actionAt" : "MIA",
         "actionCode" : "ONLOAD to LAX",
         "actionTime" : "2019-02-01T06:13:00Z"
       }, {
         "actionAt" : "MIA",
         "actionCode" : "BagTag Scan at MIA",
         "actionTime" : "2019-02-01T05:47:00Z"
       }, {
         "actionAt" : "MIA",
         "actionCode" : "Checkin at MIA",
         "actionTime" : "2019-02-01T04:38:00Z"
       } ]
   }'''
   insert_record(handle,'ticket',data1)
   insert_record(handle,'ticket.bagInfo',data2)
   insert_record(handle,'ticket.bagInfo.flightLegs',data3)
   sql_stmt_loj='SELECT * FROM ticket a LEFT OUTER JOIN ticket.bagInfo.flightLegs b ON a.ticketNo=b.ticketNo'
   print('Fetching data using Left Outer Join ')
   fetch_data(handle,sql_stmt_loj)
   sql_stmt_nt='SELECT * FROM NESTED TABLES (ticket a descendants(ticket.bagInfo.flightLegs b))'
   print('Fetching data using NESTED TABLES ')
   fetch_data(handle,sql_stmt_nt)
   sql_stmt_ij='SELECT * FROM ticket a, ticket.bagInfo.flightLegs b WHERE a.ticketNo=b.ticketNo'
   print('Fetching data using Inner Join ')
   fetch_data(handle,sql_stmt_ij)      
   if handle is not None:
      handle.close()
  
if __name__ == "__main__":
   main()
